### DISCUSSION TITLE
Focus space without disabling SIP

### DISCUSSION THREAD

**Discussion by typoon** [asmvik/yabai#803] [Q&A] [answered]
Hi,

I noticed that with SIP enabled it seems like it is not possible to move a window to another space. Amethyst seems to be able to do that, so I did some digging and looks like they use `CGSpacesInfo` for that as seen here: https://github.com/ianyh/Amethyst/blob/7cd205a1f8460d19f407acfde6aa2a33b7b0cfe7/Amethyst/Managers/WindowManager.swift#L642

Would it be possible to implement the feature to move windows to a different space in yabai using this API so that SIP can remain enabled?

Thanks!

**Comment by rockyzhang24** [comment]
I think "moving a window to another space or display" does not require SIP disabled. It is not listed on the wiki page https://github.com/koekeishiya/yabai/wiki/Disabling-System-Integrity-Protection and https://github.com/koekeishiya/yabai/issues/13#issuecomment-497950485. 

I uninstalled scriping-addition just now by `sudo yabai --uninstall-sa, killall Dock`. Then I tested "moving a window in space 1 to space 2" via `yabai -m window --space 2`, and it worked well.

**Comment by typoon** [comment]
You are completely correct. I got confused between the move window to space
and the focus space commands. It is focusing a space that does not work
without SIP disabled.

Sorry for the confusion. I guess it is safe to ignore me here then.

Any idea how Mission Control might be doing that? I would assume there
might be some private API for it that could be used?

**Comment by rockyzhang24** [comment]
I don't know the underlying techniques that Mission Control uses to do that. But I have another way to switch workspace with SIP disabled by means of the macOS Keyboard shortcuts. 

Firstly, I set the shortcuts in "System Preferences -> Keyboard -> shortcuts -> Mission Control" for workspace switching. For example `^1`, `^2`, ..., `^N` for switching to workspace 1, 2, ..., N. 

Then I create a shell script. I can pass in an argument as the target workspace id, and this script will help us switch to that workspace. The shell script is here (https://github.com/yanzhang0219/dotfiles/blob/master/.config/yabai/scripts/focus-space-SIP.sh)

Finally, in `skhdrc`, we just need to create some keybindings for switching workspaces using this script. For example, `alt - 3 : ~/.config/yabai/scripts/focus-space-SIP.sh 3 ` for switch to workspace 3.

**Reply by miguelfbrito** [reply]
The script works perfectly when running it on terminal, however it doesn't seem to work properly when combined with skhd. Any idea why is that?

**Reply by rockyzhang24** [reply]
Hello, 
Sorry for the late reply. I just now tried it, and yes it didn't work in skhd. I have no idea why it works well in terminal but not in skhd. 
Besides this, in Mission Control, I tried to change the corresponding shortcuts to `command+control+1/2/3/...` and then adjusted the applescript code in the shell script to `{control down, command down}`. Then in skhd I bind it to like `lcmd` - 3 : ~/.config/yabai/scripts/focus-space-SIP.sh 3`. It got works. So weird. Still no idea why the original doesn't work.

**Reply by jacktuck** [reply]
I've also got this issue but changing the binding to cmd + ctrl didn't help :(

**Comment by typoon** [comment]
Yeah, that is what I have been doing. Only problem I have with that is that
there doesnt seem to be a reliable way of exporting those keyboard
shortcuts from MacOS, so whenever I setup a new machine I have to recreate
those shortcuts manually.

I will google around some more and see if I can find any pointers.

**Comment by rockyzhang24** [comment]
Exactly. 
I have no idea where they are stored (probably in some plist file). What I am doing is enabling the automatic Time Machine backup so that I can set up a new machine using the latest backup in the future. 

However, at least for me, this is not the most troublesome. What bothers me the most is that there are no shortcuts in macOS to create and destroy workspaces. For workspace creation, what I can come up with is using a piece of AppleScript to simulate the mouse operation, i.e., open Mission Control, click the `+` icon on the top right corner. However, I have no idea about how to destroy a specific workspace, even a stupid one.

**Comment by hpl002** [comment]
Not to jump on the thread here, but creating a completely separate repository that describes how one can successfully mimic the same behavior as yabai without disabling SIP would be extremely useful.

Im sure many others would not only find this useful but also happily try and resolve some issues if you were to collect all of this in one place.

**Comment by rockyzhang24** [comment]
@hpl002 Thank you for your advice. I am a heavy user of yabai, and that's what I am thinking. In the Discussions, we have a section called "Ideas" and I think that's a great place to gather tips/skills. 

I learned a lot during my using and configuring yabai. I am really desired to share them. I will create a brand new thread there later to share what I've learned and come up with.

**Comment by hpl002** [comment]
> @hpl002 Thank you for your advice. I am a heavy user of yabai, and that's what I am thinking. In the Discussions, we have a section called "Ideas" and I think that's a great place to gather tips/skills.
> 
> I learned a lot during my using and configuring yabai. I am really desired to share them. I will create a brand new thread there later to share what I've learned and come up with.

Awesome! Thank you for your willingness. Looking forward to having a look 🎉

**Comment by asmvik** [comment, marked as answer]
Moved to disussions and renamed.

In short, the only way to properly focus a space without the scripting-addition is to use the shortcuts provided through mission-control. The operation of switching spaces is a combination of some lower-level functions that are a part of the SkyLight.framework and some state that is stored internally inside the Dock.app process. In fact, the entire workspaces (and mission-control) functionality of macOS is implemented inside Dock.app.

**Comment by 24unix** [comment]
There is another options besides Mission Control. I use Total Spaces for that, it brings the benefit that you can combine display to desktops.
<img width="526" alt="image" src="https://user-images.githubusercontent.com/36622631/105967084-c069ad00-6085-11eb-8061-4a98b7846c89.png">

**Reply by asmvik** [reply]
TotalSpaces also needs SIP to be disabled. It works by injecting code into Dock similar to yabai

**Reply by 24unix** [reply]
Oh, I wasn't aware that it needs SIP to be disabled, too, but it is still an alternative to Mission Control, as that is limited to 18 spaces.

**Comment by basbebe** [comment]
this isn't as elegant as the solution posted by @yanzhang0219 but this seems to work for me:

```
# move window to next space
ctrl + alt + cmd - 0x2F : yabai -m window --space next && osascript -e 'tell application "System Events" to key code 124 using control down'
ctrl + alt + cmd - 0x2B : yabai -m window --space prev && osascript -e 'tell application "System Events" to key code 123 using control down'

# move window to space and follow
ctrl + alt + cmd - 1 : yabai -m window --space 1 && osascript -e 'tell application "System Events" to key code 18 using control down'
ctrl + alt + cmd - 2 : yabai -m window --space 2 && osascript -e 'tell application "System Events" to key code 19 using control down'
ctrl + alt + cmd - 3 : yabai -m window --space 3 && osascript -e 'tell application "System Events" to key code 20 using control down'
ctrl + alt + cmd - 4 : yabai -m window --space 4 && osascript -e 'tell application "System Events" to key code 21 using control down'
ctrl + alt + cmd - 5 : yabai -m window --space 5 && osascript -e 'tell application "System Events" to key code 23 using control down'
ctrl + alt + cmd - 6 : yabai -m window --space 6 && osascript -e 'tell application "System Events" to key code 22 using control down'
ctrl + alt + cmd - 7 : yabai -m window --space 7 && osascript -e 'tell application "System Events" to key code 26 using control down'
ctrl + alt + cmd - 8 : yabai -m window --space 8 && osascript -e 'tell application "System Events" to key code 28 using control down'
ctrl + alt + cmd - 9 : yabai -m window --space 9 && osascript -e 'tell application "System Events" to key code 25 using control down'
ctrl + alt + cmd - 0 : yabai -m window --space 0 && osascript -e 'tell application "System Events" to key code 29 using control down'
```

I got asked to allow skhd usability access and then it worked, althoug not always completely reliable.
But good enough for me!

btw: what I fount out was that you **must not** keep the modifier keys pressed longer than abolutely necessary – otherwise the osescript will see those as well and won't work.

[source](https://superuser.com/a/368036/1179299)

**Reply by jacktuck** [reply]
`skhd -k "ctrl - 1"` (1-9)  works much more reliably for me

**Reply by jacktuck** [reply]
I think you can effectively replace the focus-sip script with something like this (not tested)
```
alt + cmd - 1 : yabai -m display --focus $(yabai -m query --displays --space 1 | jq '.index'); skhd -k "ctrl - 1"
alt + cmd - 2 : yabai -m display --focus $(yabai -m query --displays --space 2 | jq '.index'); skhd -k "ctrl - 2"
```

**Reply by jacktuck** [reply]
then you can do this (not sure if this is a good idea, though)
```
# move window to space
shift + alt + cmd - 1 : yabai -m window --space 1; skhd -k "alt + cmd - 1"
shift + alt + cmd - 2 : yabai -m window --space 2; skhd -k "alt + cmd - 2"
```

**Reply by ugurozturk** [reply]
Move next screen then focus next screen
```shell
shift + alt - n : yabai -m window --space next; skhd -k "ctrl + alt + shift - $(yabai -m query --spaces --space | jq '.index + 1')"
```

**Reply by ugurozturk** [reply]
Mac's Mission Control focusing screens shortcut is  = ctrl + alt + shift + [number of screen]
Pre generate 5 screen each monitor

Here is my latest configure for move next and move prev commands

```
shift + alt - p : yabai -m window --space prev & \
    space_count=$(yabai -m query --displays --display 1 | jq ".spaces | length") ; \
    index=$(yabai -m query --spaces --space | jq -r ".index") ; \
    result=$((index - 1 - (space_count - 5))) ; \
    skhd -k "ctrl + alt + shift - $result"
shift + alt - n : yabai -m window --space next & \
    space_count=$(yabai -m query --displays --display 1 | jq ".spaces | length") ; \
    index=$(yabai -m query --spaces --space | jq -r ".index") ; \
    result=$((index + 1 - (space_count - 5))) ; \
    skhd -k "ctrl + alt + shift - $result"
```

Note : Its does minuses of full screened app counts. You always should put full screened screen application to beginning of screens order
